substrate/rwlock

```diff
@@ -0,0 +1,49 @@
+// SPDX-License-Identifier: BSD-3-Clause
+#ifndef SUBSTRATE_RWLOCK
+#define SUBSTRATE_RWLOCK
+
+#if __cplusplus < 201703L
+#error "rwlock is on available on C++17 and above"
+#endif
+
+#include <substrate/internal/defs>
+
+#include <utility>
+#include <mutex>
+#include <shared_mutex>
+
+namespace substrate
+{
+	template<typename T>
+	struct rwlock final {
```

Comment:
The code style used generally puts these { on their own next line, we should be consistent about this.